src/semver.ts

```diff
@@ -0,0 +1,11 @@
+import {execSync} from 'child_process';
+
+export function satisfies(version: string, constraint: string): boolean {
+    try {
+        execSync(`/usr/local/bin/composer-semver.phar semver:match "${version}" "${constraint}" > /dev/null`);
```

Comment:
There are some edge case differences between how composer and npm resolve semantic versions (or so I'm led to believe), but I think we could use the official semver module used by npm https://www.npmjs.com/package/semver and not even require PHP here

Reply:
To clarify, I know this PR is removing it and replacing it with composer, I'm just wondering why?

Reply:
Since we do parse `composer.json`, that file could provide constraints which are mostly supported by composer but i.e. not by the NPM package.

I have not tested this yet, but composer provides some quirks:
- `^v8.1.0` would be a valid constraint
- Comma annotations are valid constraints (as outlined in #86)

There might be other quirks but since I do not want to hassle with these in this component since we should focus on detecting the correct supported version(s) based on what is supported by composer, we should use composers own logic to detect whats allowed and what not.
But thats just my opinion here.

Reply:
That first one would be a valid constraint in node aswell FYI. I did look through the docs for both npm and composer and the comma vs space was the only one I found. Which IMO is npm's mistake here as I know of other examples where a comma is perfectly valid (e.g. go).

Reply:
I have no strong opinion on this, I just would prefer not to manipulate the constraint to "make it work" 😅 
But I am open for every other suggestion. Should we report this upstream?

Reply:
Valid. I also see the case for using composer directly. It's a good argument. I just have an uncomfortable feeling about mixing services within a single container, and more than that mixing languages.

Maybe we should consider porting this to PHP instead 🤣. (some day maybe)

Reply:
TBH: I was surprised that @weierophinney wrote this in javascript in the first place. I guess that was due to the fact that github provides a library to actually interact with the GHA "API".

Reply:
I wrote it in JS for a few reasons:

- I'm reasonably proficient in it already.
- The base workflow "container" that GHA provides has JS available out-of-the-box, and includes the GHA API (as @boesing notes).
- I'd not had a chance (and still have not!) to learn TypeScript, and didn't feel I had the luxury of doing that in order to put this together (was trying to address a very real pain point we had with CI, and had limits both in the time I had available, and the amount of time Marco had then to review and help me revise).

I'm happy to see this move to TS; I think it gives us a better safety net. Unfortunately, it also means that for now, I'm not really able to review and provide meaningful feedback!

Reply:
Regarding the "why choose JS", part of the decision was _performance_. One of the requirements we put for ourselves was that generation of the matrix should be **fast** - as in, no more than 3 seconds. The only way we could get that sort of speed with PHP would be if we created a container for the action with minimal dependencies... and I wasn't sure we could do that reasonably. Today, I think we could.

Reply:
@weierophinney do you have any thoughts at an abstract level on an alternative approach to call `composer/semver` from the matrix app? The method implemented here is to include PHP in a node container and call semver via `execShell`.

Reply:
what about calling the php container via docker here instead? I'm not that familiar with how DinD works on github specifically but something like:

`execShell('docker run php /usr/local/bin/composer-semver.phar semver:match "${version}" "${constraint}" > /dev/null')`. After adding `composer-semver` to the php container instead?

Reply:
Won't work; you're operating _inside_ a container, so the container can not operate as a docker host.

As far as other suggestions... I don't have any. I used the NPM semver package originally, as I assumed it was compatible with how Composer handled semver. Knowing that they're different, we can either:

- Call on a PHP executable (as proposed here)
- Write our own TS utility that passes the same tests as composer/semver (lots more effort, plus it can go out-of-date)

It might make sense to re-do this action such that it uses a dedicated container where we use a PHP application, as we've done with the CI action and the automatic releases action.

Reply:
> Won't work; you're operating inside a container, so the container can not operate as a docker host.

Thats actually not true. It definitely works, I am just not sure if it works within GHA.
One has to mount the docker socket to the container tho.

Reply:
I am still wondering if we should just report upstream that they can't handle `,` in the constraint?
Maybe its a fixable issue for them?

Reply:
Issue opened https://github.com/npm/node-semver/issues/468, we can see what they say